src/devices/cpu/z80/t6a84.cpp

```diff
@@ -0,0 +1,227 @@
+// license:BSD-3-Clause
+// copyright-holders:QUFB
+/***************************************************************************
+
+    Toshiba T6A84, TLCS-Z80 ASSP Family
+
+***************************************************************************/
+
+#include "emu.h"
+#include "t6a84.h"
+
+#define LOG_PAGE_R (1U << 1)
+#define LOG_PAGE_W (1U << 2)
+#define LOG_MEM    (1U << 3)
+
+//#define VERBOSE (LOG_PAGE_R | LOG_PAGE_W | LOG_MEM)
+#include "logmacro.h"
+
+
+DEFINE_DEVICE_TYPE(T6A84, t6a84_device, "t6a84", "Toshiba T6A84")
+
+void t6a84_device::internal_io_map(address_map &map) const
+{
+	map.global_mask(0xff);
+	map(0xfc, 0xfc).rw(FUNC(t6a84_device::stack_page_r), FUNC(t6a84_device::stack_page_w));
+	map(0xfd, 0xfd).rw(FUNC(t6a84_device::data_page_r), FUNC(t6a84_device::data_page_w));
+	map(0xfe, 0xfe).rw(FUNC(t6a84_device::code_page_r), FUNC(t6a84_device::code_page_w));
+	map(0xff, 0xff).rw(FUNC(t6a84_device::vector_page_r), FUNC(t6a84_device::vector_page_w));
+}
+
+t6a84_device::t6a84_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	t6a84_device(mconfig, T6A84, tag, owner, clock, address_map_constructor(FUNC(t6a84_device::internal_io_map), this))
+{ }
+
+t6a84_device::t6a84_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock, address_map_constructor io_map)
+	: z80_device(mconfig, type, tag, owner, clock)
+	, m_program_space_config("program", ENDIANNESS_LITTLE, 8, 20, 0, 16, 0)
+	, m_data_space_config("data", ENDIANNESS_LITTLE, 8, 20, 0, 16, 0)
+	, m_stack_space_config("stack", ENDIANNESS_LITTLE, 8, 20, 0, 16, 0)
+	, m_io_space_config("io", ENDIANNESS_LITTLE, 8, 16, 0, io_map)
+	, m_code_page(0)
+	, m_delay_code_page(0)
+	, m_delay_code_pc(0)
+	, m_is_delay_code_page_set(false)
+	, m_prev_code_page(0)
+	, m_data_page(8)
+	, m_stack_page(8)
+	, m_vector_page(0)
+{
+	// Interrupt vectors need to be fetched and executed from their corresponding page.
+	// For simplicity, we switch pages via callbacks, instead of using a dedicated address space.
+	irqack_cb().set([this](int state) {
+		LOGMASKED(LOG_PAGE_W, "IRQ ACK %02x => %02x\n", m_code_page, m_vector_page);
+		m_prev_code_page = m_code_page;
+		m_code_page = m_vector_page;
+	});
+	reti_cb().set([this](int state) {
+		LOGMASKED(LOG_PAGE_W, "IRQ RET %02x => %02x\n", m_code_page, m_prev_code_page);
+		m_code_page = m_prev_code_page;
+	});
```

Comment:
This isn’t really how you’re supposed to use these.  They’re intended for systems that detect interrupt acknowledge and return from interrupt cycles to use.  If you use them internally, the system can’t connect anything to them.

Does the Z80 core provide a way to customise the interrupt handling for derived classes?

Reply:
The closest is `z80_daisy_chain_interface` which has callbacks for both interrupt ack and reti. But it also makes assumptions about the system architecture that I'm not sure about. We don't know if T6A84 has components using daisy-chaining, maybe we can add a separate callback for interrupt ack for now, just like I did here with `reti_cb`?